packages/x-data-grid-premium/src/components/promptControl/GridToolbarPromptControl.tsx

```diff
@@ -0,0 +1,276 @@
+  const columnExamples: Record<string, any[]> = {};
+
+  columns.forEach((column) => {
+    columnExamples[column.field] = Array.from({ length: Math.min(5, rows.length) }).map(() => {
+      const row = rows[Math.floor(Math.random() * rows.length)];
+      if (column.valueGetter) {
+        return column.valueGetter(row[column.field] as never, row, column, apiRef);
+      }
+      return row[column.field];
+    });
+  });
+
+  return columnExamples;
+}
+
+function generateContext(
+  apiRef: React.MutableRefObject<GridApiPremium>,
+  examples?: Record<string, any[]>,
+) {
+  const columns = gridColumnDefinitionsSelector(apiRef);
+  const columnsContext = columns.map((column) => ({
+    field: column.field,
+    description: column.description ?? null,
+    examples: examples?.[column.field] ?? column.unstable_examples ?? [],
+    type: column.type ?? 'string',
+    allowedOperators: column.filterOperators?.map((operator) => operator.value) ?? [],
+  }));
+
+  return `The columns are described by the following JSON:\n${JSON.stringify(columnsContext)}`;
+}
+
+type GridToolbarPromptControlProps = {
+  /**
+   * Called when the new prompt is ready to be processed.
+   * Provides the prompt and the data context and expects the grid state updates to be returned.
+   * @param {string} context The context of the prompt
+   * @param {string} query The query to process
+   * @returns {Promise<PromptResponse>} The grid state updates
+   */
+  onPrompt: (context: string, query: string) => Promise<PromptResponse>;
+  /**
+   * Allow taking couple of random cell values from each column to improve the prompt context.
+   * If allowed, samples are taken from different rows.
+   * If not allowed, the column examples are used.
+   * @default false
+   */
+  allowDataSampling?: boolean;
+  /**
+   * The BCP 47 language tag to use for the speech recognition.
+   * @default HTML lang attribute value or the user agent's language setting
+   */
+  lang?: string;
+};
+
+function GridToolbarPromptControl(props: GridToolbarPromptControlProps) {
+  const apiRef = useGridApiContext();
+  const rootProps = useGridRootProps();
+  const { onPrompt, lang, allowDataSampling = false } = props;
+
+  const [isLoading, setLoading] = React.useState(false);
+  const [error, setError] = React.useState<string | null>(null);
+  const [isRecording, setRecording] = React.useState(false);
+  const [query, setQuery] = React.useState('');
+
+  const classes = useUtilityClasses(rootProps, isRecording);
+  const examplesFromData = React.useMemo(
+    () => (allowDataSampling ? sampleData(apiRef) : undefined),
+    [apiRef, allowDataSampling],
+  );
+
+  const processPrompt = React.useCallback(() => {
+    const context = generateContext(apiRef, examplesFromData);
+    const columnsByField = gridColumnLookupSelector(apiRef);
+
+    setLoading(true);
+    setError(null);
+    apiRef.current.setState((state) => ({ ...state, rows: { ...state.rows, loading: true } }));
```

Comment:
Use API method instead?
```suggestion
    apiRef.current.setLoading(true);
```